module/extension/Director/src/bits/compare_priority.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * This file is part of the NUbots Codebase.
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2021 NUbots <nubots@nubots.net>
+ */
+
+#include <algorithm>
+#include <tuple>
+#include <vector>
+
+#include "../Director.hpp"
+
+#include "extension/Behaviour.hpp"
+
+namespace module::extension {
+
+    using ::extension::behaviour::commands::DirectorTask;
+
+    bool Director::compare_priority(const std::shared_ptr<const DirectorTask>& incumbent,
+                                    const std::shared_ptr<const DirectorTask>& challenger) {
+
+        // Compare to yourself and you get that you shouldn't change tasks
+        if (incumbent->requester_id == challenger->requester_id) {
+            return false;
+        }
+
+        // Function to get the priorities of the ancestors of this task
+        auto get_ancestor_priorities = [this](const std::shared_ptr<const DirectorTask>& task) {
+            // Loop up through the providers until we reach a point where a task was emitted by a non provider
+            std::vector<std::tuple<uint64_t, int, bool>> ancestors;
+            for (auto t = task; providers.count(t->requester_id) != 0;) {
+                auto provider = providers[t->requester_id];
+                auto& group   = groups[provider->type];
+
+                if (group.active_task != nullptr) {
+                    ancestors.emplace_back(t->requester_id, t->priority, t->optional);
+                    t = group.active_task;
+                }
+                else {
+                    throw std::runtime_error("Task has broken parentage");
+                }
+            }
+            return ancestors;
+        };
+
+        // Convenience references for the two to make the code shorter
+        const auto& a = incumbent;
+        const auto& b = challenger;
```

Comment:
I'm not sure if this is that useful, it's not like there is a lot of code. But I guess at the same time, there isn't a lot of code so it's not hard to look back and see what these mean... I'm mostly thinking because `a` and `b` are not things that can be swapped around, like in an `add(int a, int b)` function. They have specific meanings where one is the currently running task and the other is not.

Reply:
I mainly did this to make the following lines one liners. which made it easier for me to read because the function ends up less lines of code. If I change a and b back it gains about 10ish lines of code

Reply:
Ah ok. I'd suggest `c` and `i` but those are probably the worst letters to have 😆